### DISCUSSION TITLE
How can I make ramda's omit in fp-ts?

### DISCUSSION THREAD

**Discussion by ammarrik** [gcanti/fp-ts#1818] [Q&A] [answered]
A friend of mine made a function for deleting multiples values from object but types are not working. I need a help to make types work.
function is;
```js
import { pipe } from 'fp-ts/lib/function';
import { map } from 'fp-ts/Array';
import { deleteAt } from 'fp-ts/Record';

const person = { name: 'Ammar', email: 'ammar@gmail.com', dob: '12/12/2000' };

const surround = map(deleteAt);
const omit = (arr: string[], obj: object) => pipe(obj, ...surround(arr));
console.log(omit(['name', 'dob'], person));
```
and the current error while passing the surround function in pipe and using spread operator is;

``` A spread argument must either have a tuple type or be passed to a rest```

**Comment by ankrut** [comment, marked as answer]
Your code does not work because `pipe` does not support rest parameter for arrays. It is not "_passed to a rest_". In other words `pipe` is unfortunately not generic for an arbitrary number of parameters. It supports only max 20 parameters and that's why it is saying "_must [...] have a tuple type_".

To reflect a loop in functional programming [recursion is preferred](https://deepu.tech/functional-programming-in-typescript/). In your example you would need to use `reduce`. For example

```ts
import { pipe } from "fp-ts/lib/function";
import { map, reduce } from "fp-ts/Array";
import { deleteAt } from "fp-ts/Record";

// ...

const omit = (arr: string[], obj: object) =>
  pipe(
      arr,
      surround,
      reduce(obj, (reducedObj, f) => f(reducedObj))
  );
```

However, this is just a quick idea how to fix the error. The output type is still just `object`.

Please take a look at the [ `omit`](https://github.com/samhh/fp-ts-std/blob/master/src/Struct.ts#L88) method in [fp-ts-std](https://github.com/samhh/fp-ts-std). It has proper type support.

```ts
import { omit } from "fp-ts-std/Struct";

const reducedPerson = pipe(person, omit(["name", "dob"]))
```

**Reply by ammarrik** [reply]
I think fp-ts-std works well with fp-ts because the name implies that it is an extension for fp-ts. right?

**Reply by ankrut** [reply]
Yes. At least that's what it is saying:

> The missing pseudo-standard library for [fp-ts](https://gcanti.github.io/fp-ts/).

And it is official part of the fp-ts [ecosystem](https://gcanti.github.io/fp-ts/ecosystem/).
